File: src/android/ClerkPlugin.java
package com.outsystems.plugin.clerk;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.clerk.api.Clerk;
import com.clerk.api.ClerkConfigurationOptions;
import com.clerk.api.SharedSessionSyncConfig;
import com.clerk.api.models.Session;
import com.clerk.api.models.User;

/**
 * Pure Java Cordova Plugin for Clerk Authentication & Shared Session Synchronization.
 * 100% guaranteed native execution and callback resolution across all OutSystems MAB versions.
 */
public class ClerkPlugin extends CordovaPlugin {

    private boolean isInitialized = false;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
    }

    @Override
    public void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (intent != null && isInitialized) {
            try {
                Clerk.onNewIntent(intent);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if ("initialize".equals(action)) {
            this.initializeClerk(args, callbackContext);
            return true;
        } else if ("signIn".equals(action)) {
            this.signIn(callbackContext);
            return true;
        } else if ("signOut".equals(action)) {
            this.signOut(callbackContext);
            return true;
        } else if ("getToken".equals(action)) {
            this.getToken(callbackContext);
            return true;
        } else if ("getSessionState".equals(action)) {
            this.getSessionState(callbackContext);
            return true;
        }
        return false;
    }

    private void initializeClerk(JSONArray args, final CallbackContext callbackContext) {
        try {
            final String publishableKey = args.optString(0, "");
            if (publishableKey == null || publishableKey.trim().length() == 0) {
                callbackContext.error("Missing Clerk Publishable Key.");
                return;
            }

            final Context context = cordova.getActivity().getApplicationContext();

            cordova.getActivity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    try {
                        ClerkConfigurationOptions options = new ClerkConfigurationOptions(
                            false, // enableDebugMode
                            null,  // proxyUrl
                            true,  // telemetryEnabled
                            SharedSessionSyncConfig.enabled
                        );

                        Clerk.initialize(context, publishableKey, options);
                        isInitialized = true;

                        JSONObject response = new JSONObject();
                        response.put("status", "initialized");
                        response.put("sharedSessionEnabled", true);
                        response.put("publishableKey", publishableKey);

                        callbackContext.success(response);
                    } catch (Exception e) {
                        callbackContext.error("Clerk Initialization Error: " + e.getMessage());
                    }
                }
            });
        } catch (Exception e) {
            callbackContext.error("Initialize Plugin Error: " + e.getMessage());
        }
    }

    private void signIn(final CallbackContext callbackContext) {
        if (!isInitialized) {
            callbackContext.error("Clerk SDK is not initialized yet. Call initialize() first.");
            return;
        }

        final Activity activity = cordova.getActivity();
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    Clerk.getAuth().startHostedAuth(activity);
                    
                    User user = Clerk.getUser();
                    Session session = Clerk.getSession();

                    JSONObject response = new JSONObject();
                    response.put("status", "success");
                    response.put("userId", user != null ? user.getId() : "");
                    response.put("sessionId", session != null ? session.getId() : "");
                    response.put("email", (user != null && user.getPrimaryEmailAddress() != null) ? user.getPrimaryEmailAddress().getEmailAddress() : "");
                    response.put("firstName", user != null ? user.getFirstName() : "");
                    response.put("lastName", user != null ? user.getLastName() : "");

                    callbackContext.success(response);
                } catch (Exception e) {
                    callbackContext.error("Authentication Failed: " + e.getMessage());
                }
            }
        });
    }

    private void signOut(final CallbackContext callbackContext) {
        if (!isInitialized) {
            callbackContext.error("Clerk SDK is not initialized.");
            return;
        }

        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Clerk.signOut();
                    JSONObject response = new JSONObject();
                    response.put("status", "signed_out");
                    response.put("message", "User session successfully signed out.");
                    callbackContext.success(response);
                } catch (Exception e) {
                    callbackContext.error("SignOut Error: " + e.getMessage());
                }
            }
        });
    }

    private void getToken(final CallbackContext callbackContext) {
        if (!isInitialized) {
            callbackContext.error("Clerk SDK is not initialized.");
            return;
        }

        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Session session = Clerk.getSession();
                    String token = session != null ? session.getToken() : null;

                    if (token != null && token.length() > 0) {
                        JSONObject res = new JSONObject();
                        res.put("token", token);
                        callbackContext.success(res);
                    } else {
                        callbackContext.error("No active session token found.");
                    }
                } catch (Exception e) {
                    callbackContext.error("Token Retrieval Error: " + e.getMessage());
                }
            }
        });
    }

    private void getSessionState(final CallbackContext callbackContext) {
        try {
            if (!isInitialized) {
                JSONObject response = new JSONObject();
                response.put("isAuthenticated", false);
                response.put("sessionId", "");
                response.put("userId", "");
                response.put("email", "");
                response.put("sharedSessionSynced", false);
                callbackContext.success(response);
                return;
            }

            Session session = Clerk.getSession();
            User user = Clerk.getUser();
            boolean isAuthenticated = session != null && session.isActive();

            JSONObject response = new JSONObject();
            response.put("isAuthenticated", isAuthenticated);
            response.put("sessionId", session != null ? session.getId() : "");
            response.put("userId", user != null ? user.getId() : "");
            response.put("email", (user != null && user.getPrimaryEmailAddress() != null) ? user.getPrimaryEmailAddress().getEmailAddress() : "");
            response.put("sharedSessionSynced", isAuthenticated);

            callbackContext.success(response);
        } catch (Exception e) {
            callbackContext.error("Session Check Error: " + e.getMessage());
        }
    }
}

File: src/main/android/ClerkPlugin.kt
package com.outsystems.plugin.clerk

import android.content.Intent
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaInterface
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.CordovaWebView
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/**
 * Clerk Cordova Plugin written in Kotlin for OutSystems Mobile App Builder (MAB).
 * Fully compatible with Clerk Android SDK and Shared Session Synchronization.
 */
class ClerkPlugin : CordovaPlugin() {

    override fun initialize(cordova: CordovaInterface, webView: CordovaWebView) {
        super.initialize(cordova, webView)
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        // Pass incoming deep link intent (clerk://...) to Clerk SDK
        ClerkBridge.handleIntent(intent)
    }

    @Throws(JSONException::class)
    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        return when (action) {
            "initialize" -> {
                initializeClerk(args, callbackContext)
                true
            }
            "signIn" -> {
                signIn(callbackContext)
                true
            }
            "signOut" -> {
                signOut(callbackContext)
                true
            }
            "getToken" -> {
                getToken(callbackContext)
                true
            }
            "getSessionState" -> {
                getSessionState(callbackContext)
                true
            }
            else -> false
        }
    }

    private fun initializeClerk(args: JSONArray, callbackContext: CallbackContext) {
        try {
            val publishableKey = if (!args.isNull(0)) args.getString(0) else ""
            if (publishableKey.isEmpty()) {
                callbackContext.error("Missing Clerk Publishable Key.")
                return
            }

            cordova.activity.runOnUiThread {
                ClerkBridge.initialize(
                    cordova.activity.applicationContext,
                    publishableKey,
                    object : ClerkBridge.BridgeCallback {
                        override fun onSuccess(result: JSONObject) {
                            callbackContext.success(result)
                        }

                        override fun onError(errorMessage: String) {
                            callbackContext.error(errorMessage)
                        }
                    }
                )
            }
        } catch (e: Exception) {
            callbackContext.error("Initialize Plugin Error: ${e.localizedMessage}")
        }
    }

    private fun signIn(callbackContext: CallbackContext) {
        cordova.activity.runOnUiThread {
            ClerkBridge.startHostedAuth(
                cordova.activity,
                object : ClerkBridge.BridgeCallback {
                    override fun onSuccess(result: JSONObject) {
                        callbackContext.success(result)
                    }

                    override fun onError(errorMessage: String) {
                        callbackContext.error(errorMessage)
                    }
                }
            )
        }
    }

    private fun signOut(callbackContext: CallbackContext) {
        ClerkBridge.signOut(object : ClerkBridge.BridgeCallback {
            override fun onSuccess(result: JSONObject) {
                callbackContext.success(result)
            }

            override fun onError(errorMessage: String) {
                callbackContext.error(errorMessage)
            }
        })
    }

    private fun getToken(callbackContext: CallbackContext) {
        ClerkBridge.getSessionToken(object : ClerkBridge.TokenCallback {
            override fun onSuccess(jwtToken: String) {
                try {
                    val res = JSONObject().apply {
                        put("token", jwtToken)
                    }
                    callbackContext.success(res)
                } catch (e: Exception) {
                    callbackContext.success(jwtToken)
                }
            }

            override fun onError(errorMessage: String) {
                callbackContext.error(errorMessage)
            }
        })
    }

    private fun getSessionState(callbackContext: CallbackContext) {
        ClerkBridge.getSessionState(object : ClerkBridge.BridgeCallback {
            override fun onSuccess(result: JSONObject) {
                callbackContext.success(result)
            }

            override fun onError(errorMessage: String) {
                callbackContext.error(errorMessage)
            }
        })
    }
}
